src/include/catalog/catalog_accessor.h

```diff
@@ -0,0 +1,476 @@
+#pragma once
+
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "catalog/catalog_defs.h"
+#include "catalog/default_value.h"
+#include "catalog/index_key_schema.h"
+#include "catalog/schema.h"
+#include "storage/index/index.h"
+#include "storage/sql_table.h"
+#include "type/type_id.h"
+
+namespace terrier::catalog {
+class Catalog;
+
+/**
+ * A stateful wrapper around the catalog that provides the primary mechanisms
+ * for the execution engine to interface with the catalog.  The execution engine
+ * should not access the catalog object or the underlying tables directly
+ * because that will lock us into the underlying implementation (currently
+ * PostgreSQL catalog tables) and makes any future changes or optimizations
+ * more difficult.
+ *
+ * Additionally, the catalog is not responsible for creating or managing
+ * anything at the storage layer (except for whatever backs the catalog data).
+ * In other words, 'CreateTable' only creates an entry inside of the catalog
+ * that gets assigned an OID.  It does not call out to the constructor of
+ * 'SqlTable' which is left to the execution engine.  This separation of
+ * responsibilities allows for certain policy choices (lazy table instantiation)
+ * as well as reinforces the design decision that the catalog is responsible
+ * only for managing metadata and not the lifecycle of storage objects.
+ *
+ * TODO(John):  With this interface in-place, we should probably make the
+ * constructors/destructors for Schema and Column private with the catalog
+ * classes as the only friends.  However, this may break code (especially
+ * tests).
+ */
+class CatalogAccessor {
+ public:
+  /**
+   * This class decouples the definition of a column from its existence in a
+   * schema.  This allows us to delegate all OID assignments to the catalog
+   * (rather than requiring the caller to reason about them) and allows us to
+   * treat Schema and Column objects as immutable outside of the catalog.
+   */
+  class ColumnDefinition {
```

Comment:
Should we move `ColumnDefinition` and `IndexKeyDefinition` into standalone classes? We may want to reuse these in other parts of the system (e.g., CreatePlanNodes).

Reply:
That probably makes sense.  Should they still live in the CatalogAccessor file or put them as standalone classes in their respective “schema” header files?